src/file_loader.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2013-2018, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use glib::object::IsA;
+use glib::translate::*;
+use gtk_source_sys;
+use std::boxed::Box;
+use std::ptr;
+use FileLoader;
+
+impl FileLoader {
+    pub fn load_async<
+        P: IsA<gio::Cancellable>,
+        Q: 'static + Fn(i64, i64) + Send,
+        R: 'static + FnOnce(Result<bool, glib::Error>) + Send,
```

Comment:
The `bool` is redundant here: it's always `true` or otherwise you would've gotten an `glib::Error`

```suggestion
        R: 'static + FnOnce(Result<(), glib::Error>) + Send,
```

Reply:
I thought the same initially, but found accidentally during testing that it *is* possible to have a 'false' returned but no `glib::Error`. The file was read, the progress callbacks were called, and everything looked normal, but at the end of the operation the buffer was empty. The problem was caused by constructing the `FileLoader` with a `gio::File` that was allowed to go out of scope before the call to `load_async`.  Yes, a dumb error, but is it something we would want to alert the user to?  And if so, perhaps it would be better to ditch the bool anyway, and create instead a `glib::Error` with a suitable message?  It certainly puzzled me for a while!   :)

Reply:
That sounds like a bug in the sourceview C library. Can you report it there?

Reply:
For the time being you could convert `false` without error into `glib::Error::new(gio::Error::Failed, "Unexpected failure")` I guess.

Reply:
Have tried to reproduce this bug in C, without success. Unreffing the GFile, one or more times, either works completely or fails completely. The Rust problem must depend on exactly what happens when the file is dropped.

Reply:
Can you provide a testcase for reproducing it?

Reply:
This fails with the error in question, though the behaviour is slightly different, in that the progress callback does not get called. That may well be because of the changes to `load_async` that have been made since the original version.
https://gist.github.com/brackleian/4788ad5a1b69ed487bba46548cb2a6a7

Reply:
That's completely intentional from looking at the C code. It goes out of its way to only store a weak reference to the `File` internally and checks if it's `NULL` everywhere in case it disappeared in the meantime. I have no idea why, it would be simpler not to do that so there must be a reason for it :) It's also not documented anywhere.